File: code/zato-web-admin/src/zato/admin/web/views/vendors/keysight_vision.py
# -*- coding: utf-8 -*-

"""
Copyright (C) 2023, Zato Source s.r.o. https://zato.io

Licensed under AGPLv3, see LICENSE.txt for terms and conditions.
"""

# Zato
from zato.admin.web.forms import ChangePasswordForm
from zato.admin.web.forms.vendors.keysight_vision import CreateForm, EditForm
from zato.admin.web.views import change_password as _change_password, CreateEdit, Delete as _Delete, Index as _Index, \
    method_allowed, ping_connection
from zato.common.api import Wrapper_Type
from zato.common.model.keysight_ import KeysightVisionConfigObject

# ################################################################################################################################
# ################################################################################################################################

if 0:
    from zato.common.typing_ import any_, strdict

# ################################################################################################################################
# ################################################################################################################################

class Index(_Index):
    method_allowed = 'GET'
    url_name = 'vendors-keysight-vision'
    template = 'zato/vendors/keysight-vision.html'
    service_name = 'zato.generic.rest-wrapper.get-list'
    output_class = KeysightVisionConfigObject
    paginate = True
    wrapper_type = Wrapper_Type.Keysight_Vision

    class SimpleIO(_Index.SimpleIO):
        input_required = 'cluster_id',
        output_required = 'id', 'name', 'is_active', 'host', 'username', 'sec_tls_ca_cert_id'
        output_repeated = True

# ################################################################################################################################

    def handle(self) -> 'strdict':
        return {
            'create_form': CreateForm(req=self.req),
            'edit_form': EditForm(req=self.req, prefix='edit'),
            'change_password_form': ChangePasswordForm()
        }

# ################################################################################################################################
# ################################################################################################################################

class _CreateEdit(CreateEdit):
    method_allowed = 'POST'

    class SimpleIO(CreateEdit.SimpleIO):
        input_required = 'id', 'name', 'is_active', 'host', 'username', 'sec_tls_ca_cert_id', 'password'
        output_required = 'id', 'name'

# ################################################################################################################################

    def populate_initial_input_dict(self, initial_input_dict:'strdict') -> 'None':
        initial_input_dict['wrapper_type'] = Wrapper_Type.Keysight_Vision

# ################################################################################################################################

    def success_message(self, item:'any_') -> 'str':
        return f'Successfully {self.verb} Keysight Vision Series connection `{item.name}`'

# ################################################################################################################################
# ################################################################################################################################

class Create(_CreateEdit):
    url_name = 'vendors-keysight-vision-create'
    service_name = 'zato.generic.rest-wrapper.create'

# ################################################################################################################################
# ################################################################################################################################

class Edit(_CreateEdit):
    url_name = 'vendors-keysight-vision-edit'
    form_prefix = 'edit-'
    service_name = 'zato.generic.rest-wrapper.edit'

# ################################################################################################################################
# ################################################################################################################################

class Delete(_Delete):
    url_name = 'vendors-keysight-vision-delete'
    error_message = 'Could not delete Keysight Vision Series connection'
    service_name = 'zato.generic.rest-wrapper.delete'

# ################################################################################################################################
# ################################################################################################################################

@method_allowed('POST')
def change_password(req:'any_') -> 'any_':
    return _change_password(req, 'zato.generic.rest-wrapper.change-password', success_msg='Password updated')

# ################################################################################################################################

@method_allowed('POST')
def ping(req:'any_', id:'int', cluster_id:'int') -> 'any_':
    service = 'zato.generic.rest-wrapper.ping'
    ping_path = '/api/users'
    return ping_connection(req, service, id, 'Keysight Vision Series connection', ping_path=ping_path)

# ################################################################################################################################

File: code/zato-web-admin/src/zato/admin/static/js/vendors/keysight-vision.js

// ///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

$.fn.zato.data_table.KeysightVision = new Class({
    toString: function() {
        var s = '<KeysightVision id:{0} name:{1} is_active:{2}>';
        return String.format(s, this.id ? this.id : '(none)',
                                this.name ? this.name : '(none)',
                                this.is_active ? this.is_active : '(none)');
    }
});

// ///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

$(document).ready(function() {
    $('#data-table').tablesorter();
    $.fn.zato.data_table.class_ = $.fn.zato.data_table.KeysightVision;
    $.fn.zato.data_table.new_row_func = $.fn.zato.vendors.keysight.vision.data_table.new_row;
    $.fn.zato.data_table.parse();
    $.fn.zato.data_table.setup_forms([
        'name',
        'host',
        'username',
        'password',
        'sec_tls_ca_cert_id',
    ]);
})

// ///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

$.fn.zato.vendors.keysight.vision.create = function() {
    $.fn.zato.data_table._create_edit('create', 'Create a new Keysight Vision Series connection', null);
}

// ///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

$.fn.zato.vendors.keysight.vision.edit = function(id) {
    $.fn.zato.data_table._create_edit('edit', 'Update the Keysight Vision Series connection', id);
}

// ///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

$.fn.zato.vendors.keysight.vision.data_table.new_row = function(item, data, include_tr) {
    let row = '';

    if(include_tr) {
        row += String.format("<tr id='tr_{0}' class='updated'>", item.id);
    }

    let is_active = item.is_active == true;

    row += "<td class='numbering'>&nbsp;</td>";
    row += "<td class='impexp'><input type='checkbox' /></td>";

    // 1
    row += String.format('<td>{0}</td>', item.name);
    row += String.format('<td>{0}</td>', is_active ? 'Yes' : 'No');
    row += String.format('<td>{0}</td>', item.host);

    // 2
    row += String.format("<td>{0}</td>", item.username);
    row += String.format("<td class='ignore'>{0}</td>", item.sec_tls_ca_cert_id);

    // 3
    row += String.format('<td>{0}</td>',
        String.format("<a href=\"javascript:$.fn.zato.data_table.change_password('{0}', 'Change password')\">Change password</a>", item.id));
    row += String.format('<td>{0}</td>', String.format("<a href=\"javascript:$.fn.zato.vendors.keysight.vision.edit('{0}')\">Edit</a>", item.id));
    row += String.format('<td>{0}</td>', String.format("<a href='javascript:$.fn.zato.vendors.keysight.vision.delete_({0});'>Delete</a>", item.id));
    row += String.format('<td>{0}</td>', String.format("<a href='javascript:$.fn.zato.data_table.ping({0});'>Ping</a>", item.id));

    // 4
    row += String.format("<td class='ignore item_id_{0}'>{0}</td>", item.id);
    row += String.format("<td class='ignore'>{0}</td>", item.is_active);

    if(include_tr) {
        row += '</tr>';
    }

    return row;
}

// ///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

$.fn.zato.vendors.keysight.vision.delete_ = function(id) {
    $.fn.zato.data_table.delete_(id, 'td.item_id_',
        'Keysight Vision Series connection `{0}` deleted',
        'Are you sure you want to delete Keysight Vision Series connection `{0}`?',
        true);
}

// ///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
